### PR TITLE
Batch Write: Optimize Key Partition

### PR DESCRIPTION
**Pull request by birdstorm**
@zhexuany 

A optimized version of key partition. Use binary search instead of linear search.

### CODE REVIEW

File: core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala [hunk 1]
```diff
@@ -916,7 +919,7 @@ case class WrappedEncodedRow(row: TiRow,
 
 // SimplifiedWrappedEncodedRow is used at last stage if write process.
```
Comment (marsishandsome) [inline core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala:920]:
```suggestion
// EncodedKVPair is used at last stage if write process.
```

File: core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala [hunk 2]
```diff
@@ -882,20 +872,33 @@ class TiBatchWrite(@transient val df: DataFrame,
...
+class TiRegionPartitioner(regions: util.List[TiRegion], writeConcurrency: Int) extends Partitioner {
+  def binarySearch(key: Key): Int = {
+    if (regions.get(0).contains(key)) {
+      return 0
+    }
+    var l = 0
+    var r = regions.size()
+    while (l < r) {
+      val mid = l + (r - l) / 2
+      val region = regions.get(mid)
```
Comment (marsishandsome) [inline core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala:884]: add following code?
```
if(region.contains(key)) 
    return mid
```

Comment (birdstorm) [reply core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala:884]: no need to return mid.

**Pull request merged**